### DISCUSSION TITLE
linter: fix codebase for same rule in different plugin/scopes

### DISCUSSION THREAD

**Discussion by Sysix** [oxc-project/oxc#7242] [General]
## Background

In the `eslint` world there are similiar rule (names) in different plugins. As an example:

- `no-unused-vars` and `@typescript-eslint/no-unused-vars`
- `vitest/no-identical-title` and `jest/no-identical-title`

## Problems

Because `oxlint` support them only in one scope (`vitest` in `jest` and `@typescript-eslint` in `eslint`),
there are some problems.

### Documentation

- https://github.com/oxc-project/oxc/pull/3779 (and some follow up)
- https://github.com/oxc-project/oxc/pull/6667

### Implementation with config

- https://github.com/oxc-project/oxc/issues/7233
- https://github.com/oxc-project/oxc/issues/7240

### Integration with eslint-plugin-oxlint

in this project we need to keep track for the configs and the new method `buildFromOxlintConfigFile`

- https://github.com/oxc-project/eslint-plugin-oxlint/issues/187
- https://github.com/oxc-project/eslint-plugin-oxlint/pull/165

## Discussion

I want to have a simple way to parse the rules for all specific use cases. 
In discord there is also a `import/no-namespace` upcoming  rule wich conflicts with `@typescript-eslint`, 
see here https://discord.com/channels/1079625926024900739/1080712072012238858/1304749400383291422
This problem crops up more often, the more plugins we are supporting.

**Comment by nrayburn-tech** [comment]
I agree that this is an issue and I would like to see it addressed somehow.  As you mention, the more plugins that are supported the more often this is an issue.  I expect it will continue to get worse, especially when custom plugins become available.

**Comment by Boshen** [comment]
It was an oversight to not allow identical names, let's treat it as a bug and fix accordingly.

**Comment by pumano** [comment]
Would love if it will be fixed, because typescript/no-namespace and import/no-namespace can be used too without changing name

**Comment by pumano** [comment]
Looks like #7082 related to that problem too.

@Sysix @nrayburn-tech how we can fix it?

**Reply by Sysix** [reply]
I do not know how the perfect solution looks like :)

My first guess is the modifiy the `rules.rs`:
- allow exporting a imported rule multiple times with different names
- renaming all `declare_all_lint_rules` rules with plugin scope
- add missing rules from same scope (vitest to jest, ts to eslint)

But this is too big of a problem for me, I do know the architecture of oxc.   
Someone from the core team should look into this:)

**Reply by pumano** [reply]
@camchenry @camc314 could you help us with that solution or guide us, how we can do it?

**Reply by camc314** [reply]
thoughts on the following:
when using the cli, rules can either be enabled with their name
e.g no-non-null-assertions 
OR their scoped name @ typescript/no-non-null-assertions

?

will try to look at this tomorrow

**Reply by pumano** [reply]
@camc314 any news on this? Looks like we need to fix naming problem.

Maybe add forced scope to all rules with exception to eslint rules?

If we implement it, we got proper working for rules below, without naming collisions:
```json
"no-unused-vars": "error", // or off
"@typescript-eslint/no-unused-vars": "error", 
"@typescript-eslint/no-namespace": "error",
"import/no-namespace": "error",
"vitest/no-identical-title": "error", 
"jest/no-identical-title": "error"
```

**Reply by camc314** [reply]
afraid not. there's been a lack of time.

**Reply by camc314** [reply]
https://github.com/oxc-project/oxc/pull/7496

**Comment by Sysix** [comment]
@camc314 & @Boshen what are the next plans to support rules, which extends some other rules?

See linked issues + a new found: https://github.com/oxc-project/eslint-plugin-oxlint/issues/263

## I tried to collect all known rules for oxc project:

### Rules which are implemented in `eslint` and extended in `@typescript-eslint`:

_Note: This are ALL rules, we do not support all of them because some are cosmetic and should be implemented in oxc_prettier_

-  'block-spacing'
-  'brace-style'
-  'class-methods-use-this'
-  'comma-dangle'
-  'comma-spacing'
-  'default-param-last'
-  'func-call-spacing'
-  'indent'
-  'init-declarations'
-  'key-spacing'
-  'keyword-spacing'
-  'lines-around-comment'
-  'lines-between-class-members'
-  'max-params'
-  'no-array-constructor'
-  'no-dupe-class-members'
-  'no-empty-function'
-  'no-extra-parens'
-  'no-extra-semi'
-  'no-invalid-this'
-  'no-loop-func'
-  'no-loss-of-precision'
-  'no-magic-numbers'
-  'no-redeclare'
-  'no-restricted-imports'
-  'no-shadow'
-  'no-unused-expressions'
-  'no-unused-vars'
-  'no-use-before-define'
-  'no-useless-constructor'
-  'object-curly-spacing'
-  'padding-line-between-statements'
-  'quotes'
-  'semi'
-  'space-before-blocks'
-  'space-before-function-paren'
-  'space-infix-ops'

### Rules which are implemented in `jest` but also supported for `vitest`:

_Note: This CAN BE all supported rules by oxc_
_I do not know if they extend are just in another plugin_

-  'consistent-test-it'
-  'expect-expect'
-  'no-alias-methods'
-  'no-conditional-expect'
-  'no-conditional-in-test'
-  'no-commented-out-tests'
-  'no-disabled-tests'
-  'no-focused-tests'
-  'no-identical-title'
-  'no-restricted-jest-methods'
-  'no-test-prefixes'
-  'prefer-hooks-in-order'
-  'valid-describe-callback'
-  'valid-expect'

### Rules which are implemented in `eslint` and and extended in `unicorn`:

_Note: This are FOUND rules supported by oxc_
- `prefer-spread`

**Reply by Boshen** [reply]
I think we should mark them as implemented in both eslint plugins.

If things start getting annoying, we may need to add some meta data to these rules.

For oxlint plugin, we may want to add an API in `oxlint` to get the number of rules with full metadata instead of parsing `rules.rs` (correct me if I am wrong).

**Reply by Sysix** [reply]
the eslint plugin `eslint-plugin-oxlint` does not parse the `rules.rs`.
Instead it looks recursivly into the rules directory and getting all informations (category: correctness, style, .., plugin-name: folder)

I am already getting annoyed because of the 3 big parts:

- Documentation in oxlint
- Detection in `oxlintrc.json` (maybe even oxlint cli, did not test it)
- Remapping in `eslint-plugin-oxlint`

**Comment by Sysix** [comment]
Closing in favour of #7622, #7574 and #7567

the rest should be `eslint-plugin-oxlint` problems and can be fixed in the own repo

**Comment by Sysix** [comment]
@camc314 Sorry to bother you, I did not want to spam in this PR https://github.com/oxc-project/oxc/pull/8149#issuecomment-2564090258

The starting discoussion will apply to the PR too. We need to sync the docs and the matching inside `eslint-plugin-oxlint`.
I still want a better solution <3

**Reply by camc314** [reply]
no worries.

do you have an example of where the docs are incorrect?
i'm not familiar with eslint-plugin-oxlint, mind pointing me to how it uses it?

or what changes you need to get it working?

**Reply by Sysix** [reply]
we needed to implement the sync for the docs in:

https://github.com/oxc-project/oxc/pull/3779
https://github.com/oxc-project/oxc/pull/6667

Now with https://github.com/oxc-project/oxc/pull/8149 we need to do the same.

For the `eslint-plugin-oxlint` problem:
I already created an PR to use the new format in #7574, you can see here:
https://github.com/oxc-project/eslint-plugin-oxlint/pull/287

I still need to remap the rules here:
https://github.com/oxc-project/eslint-plugin-oxlint/blob/build-use-oxlint-rules-command/scripts/traverse-rules.ts#L60-L79

and for the ingration tests with oxlint:
https://github.com/oxc-project/eslint-plugin-oxlint/blob/build-use-oxlint-rules-command/src/build-from-oxlint-config.spec.ts#L419-L433
